### ISSUE TITLE
Broadcasting syntax

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#39]
cc @edubart.

As mentionned in the [README](https://github.com/mratsim/Arraymancer/blob/f35e677cc74ec3e2f2f81b91104ce0c3317dbe84/README.md#broadcasting), current broadcasting syntax is not ideal.

![](https://camo.githubusercontent.com/5d3ddabc0be54c3acced709cc71ccb5aec3a8a06/68747470733a2f2f73636970792e6769746875622e696f2f6f6c642d77696b692f70616765732f696d616765303034646539652e676966)

```Nim
let j = [0, 10, 20, 30].toTensor(Cpu).reshape(4,1)
let k = [0, 1, 2].toTensor(Cpu).reshape(1,3)

echo j.bc([4,3]) + k.bc([4,3])
# Tensor of shape 4x3 of type "int" on backend "Cpu"
# |0      1       2|
# |10     11      12|
# |20     21      22|
# |30     31      32|
```

Numpy syntax would be `j + k`

This is great as it is concise, however it is probably the source of lots of bug-hunting for new Numpy users when broadcasting is not wanted. The bad thing is that the error is silent.

For this reason, I would like to default to mathematical expectations (i.e. "Incompatible shape error"), and provide an opt-in broadcasting convenience like so:

```Nim
j.bc + k.bc
```
### Implementation
A term-rewriting macro can `bc` in operations and broadcast one or both arguments to compatible shapes

**Comment by edubart** [comment]
An idea:

Implicit broadcasting like numpy is useful to type fast and with ease, however there is the case of misuse by the user, an option would be instead of doing one or other, do both, and let this **be the user choice** by declaring all the "dangerous" operators in a separate syntax module.

I will name the two syntaxes *math safe* and *math unsafe*, in the *math safe* mode only math consistent operations are declared, while in *math unsafe* both worlds are allowed. All the code in the library core would be done in math safe mode.

Example of a math safe function applied to a Tensor:

```Nim
proc sigmoid[T](x: Tensor[T]): Tensor[T] =
  result = 1.0f.bc |/| (1.0f.bc + exp(-x))
```

Now in math unsafe mode user would have to import an module with all the unsafe stuff.

```Nim
import arraymancer.unsafe_math
proc sigmoid[T](x: T): T =
  result = 1.0f / (1.0f + exp(-x))
```

Note now that this way **the function can accept now both number or tensor types**, when a tensor is passed all operations are done element-wise, so the code can be reused for both tensor and numbers. In deep learning many activation functions are done element-wise, so allowing this would be a nice.

The fallowing table would be a suggestion of the operations syntax in this idea:

|  |  math safe  |  math unsafe  |
|---|---|---|
|  matrix multiplication of compatible shapes |   a %*%  b |  a .* b |
|  matrix multiplication of incompatible shapes |   N/A |  a .* b |
|  dot product of compatible shapes |  a .* b |  a .* b |
|  dot product of incompatible shapes |  N/A |  a .* b |
|  element-wise addition of compatible shapes |  a + b |  a + b |
|  element-wise addition of incompatible shapes |  N/A |  a + b |
|  element-wise subtraction of compatible shapes |  a - b |  a - b |
|  element-wise subtraction of incompatible shapes |  N/A |  a - b |
|  element-wise multiplication of compatible shapes |  a \|*\| b |  a * b |
|  element-wise multiplication of incompatible shapes |  N/A |  a * b |
|  element-wise division of compatible shapes |  a \|/\| b |  a / b |
|  element-wise division of incompatible shapes |  N/A|  a / b |

*Where we have N/A explicit broadcasting would have to be used in one of the tensors using the .bc method*

The only syntax that would have to be changed right now is the matrix multiplication, because both modes could not have conflicts in the syntax while having `*` for element-wise in the unsafe math syntax is useful for the example above.

This math unsafe mode would be done in separated file isolated from the rest of the framework.

**Comment by mratsim** [comment]
I think that would be too much symbols to learn compared to just a `bc` "tag" after broadcasted tensors.

Note, I just checked Julia for some inspiration and element-wise/normal matrix multiplication is really a mess across languages/lib:

| Language/lib      | Normal matmul | element-wise  matmul (Hadamard) | vec-vec dot product | mat-vec multiplication|
| ------------- | ---------------------------- | --- | --- | --- |
| Arraymancer  | A * B | \|*\| | A * B | A * B |
| neo/linalg  | A * B | \|*\| | A * B | A * B |
| Julia  | A * B | .* | dot(A, B) | A * B |
| Numpy ndarray| np.dot(A, B) or np.matmul(A, B) or A @ B| np.multiply(A, B) or A * B | np.dot(A, B) or np.inner(A, B) | np.dot(A, B) |
| R | A %*% B | A * B | A %*% B or dot(A, B)| A %*% B |
| Tensorflow | tf.matmul(A, B) or A @ B | tf.multiply(A, B) | tf.matmul(a, b, transpose_a=False, transpose_b=True) or tf.tensordot(a, b, 1) or tf.einsum('i,i->', x, y) | same reshape/transpose/einsum shenanigans as vec-vec|
| Torch/PyTorch | torch.mm(A,B) or torch.matmul(A,B) | torch.cmul(A, B) | torch.dot(A, B) or torch.matmul(A, B) | torch.mv(A, B) or torch.dot(A, B)
| Theano | theano.tensor.dot(A, B) | A * B | dot(A, B) or vdot(A, B) ?| dot(A, B) or tensordot(A,B) ? |
| Common math |

I'd rather not add to that.

**Comment by mratsim** [comment]
After reading about Julia [Devectorize](https://github.com/lindahua/Devectorize.jl), I'm planning of adding a template for element-wise operation fusion mentioned in https://github.com/mratsim/Arraymancer/issues/31.

The use would be instead of this tentative broadcasting sigmoid version:
(bc shape arg is automatically computed by a term-rewriting macro)
```Nim
proc sigmoid[T](x: Tensor[T]): Tensor[T] =
  result = 1.0f.bc |/| (1.0f.bc + exp(-x))
```

we would have this:
```Nim
proc sigmoid[T](x: Tensor[T]): Tensor[T] =
  result = newTensor(x.shape)
  elementwise:
    result = 1.0 / (1.0 + exp(-x))
```
The number of operation fused would be arbitrarily long, and it should be trivial to emit custom CUDA kernel operator for fusion on Cuda

```Nim
elementwise:
  let out_of_place = alpha * A + exp(B) + beta * sin(C)
  in_place += sin(A) + exp(A + 1) + C
```

**Comment by mratsim** [comment]
I've been mulling over this and I think the .bc tag to signify broadcasting is still too noisy. I will implement Julia operators which is actually what you suggested. Normal math operations have the normal symbol, unsafe/broadcasted/elementwise operations are all prefixed with a dot like `.*`, `.+`, `./` (they actually go beyond that and you can do sin.(x)).

In the end we will have this:

|  |  math safe  |  math unsafe  |
|---|---|---|
|  matrix multiplication of compatible shapes |   a *  b |  N/A |
|  matrix multiplication of incompatible shapes |   N/A |  N/A |
|  dot product of compatible shapes |  dot(a,b) |  N/A |
|  dot product of incompatible shapes |  N/A |  N/A |
|  element-wise addition of compatible shapes |  a + b |  a .+ b |
|  element-wise addition of incompatible shapes |  N/A |  a .+ b |
|  element-wise subtraction of compatible shapes |  a - b |  a .- b |
|  element-wise subtraction of incompatible shapes |  N/A |  a .- b |
|  element-wise multiplication of compatible shapes | a * B if a scalar |  a .* b |
|  element-wise multiplication of incompatible shapes | a * B if a scalar |  a .* b |
|  element-wise division of compatible shapes |  A / b if b is scalar|  a ./ b |
|  element-wise division of incompatible shapes |  N/A|  a ./ b |

This way sigmoid can be done like this.
```Nim
1 ./ (1 .+ exp(-x))
```

I will still implement the `elementwise` template/macro later so that we access each index only once (loop fusion) and avoid temporary allocations.

**Comment by mratsim** [comment]
Note: One unfortunate consequence is breaking change for the dot product operator `.*` but it will be catched by the compiler anyway since the new `.*` won't return a scalar but a Tensor instead

**Comment by edubart** [comment]
The last suggestion looks good for me, the syntax you suggested is the same syntax in Matlab/Octave which a lot of mathematicians use, and I'm familiar with, I would vote for that.

**Comment by mratsim** [comment]
Some very interesting notes from the dev who introduced broadcasting to PyTorch: https://github.com/gchanan/pytorch/wiki/Broadcasting-Notes

**Comment by mratsim** [comment]
Closed by https://github.com/mratsim/Arraymancer/commit/ce48f6f18898c4f5352fdd2924eb0df02f0c07ba and https://github.com/mratsim/Arraymancer/commit/03e7afe614910a8551d8b56fa8f15e85160cf97f
